### PR TITLE
TEST/GTEST: Add sanity tests

### PR DESCRIPTION
**Pull request by tvegas1**
## What
Adds sanity userspace tests, starting from porting existing C tests.

## Why ?
We want to enable automated sanities.

## How ?
This includes 3 steps:
- move existing tests to `test/share` directory
- add gtest infra
- add/port tests

Tested:
- compile with and without C++14
- ran `make check`
- ran `cd test/share && ./run.sh`
- ran `./test/gtest/gtest`.

### CODE REVIEW

**Comment by brminich** [comment]
@Artemy-Mellanox

File: test/gtest/test_basic.cc [hunk 1]
```diff
@@ -5,10 +5,285 @@
...
+TEST(sanity, normal_exit) {
+  EXPECT_EXIT({ exit(0); }, ::testing::ExitedWithCode(0), "");
+}
+
+static void some_assert_failure() { exit(0); }
+
+TEST(sanity, assert_exit) {
+  EXPECT_EXIT(some_assert_failure(), ::testing::ExitedWithCode(0), "");
+}
+
```
Comment (brminich) [inline test/gtest/test_basic.cc]: do we really want to test gtest infra? if you think we need that it's better to move to a separate file

File: test/gtest/test_basic.cc [hunk 2]
```diff
@@ -5,10 +5,285 @@
...
+  if (!pid) {
+    EXPECT_TRUE(true);
+    exit(::testing::Test::HasFailure());
+  } else {
+    waitpid(pid, &status, 0);
+    ASSERT_EQ(0, status);
+  }
+}
+
```
Comment (brminich) [inline test/gtest/test_basic.cc]: - why not using macros for inspecting `status` (like `WEXITSTATUS`)?
- the code is quite similar, can define a common func (or aux class member) to reduce duplication

File: test/gtest/test_basic.cc [hunk 3]
```diff
@@ -25,5 +300,14 @@ TEST_F(test_basic, 64pages) {
...
+  auto result = pattern_check(att_ptr.get(), size, seed);
+  auto expect = std::make_pair<bool, int>(true, 0);
+  ASSERT_EQ(expect.first, result.first);
+}
+
+int main(int argc, char** argv) {
+  testing::InitGoogleTest(&argc, argv);
+  GTEST_FLAG_SET(death_test_style, "threadsafe");
+  return RUN_ALL_TESTS();
```
Comment (brminich) [inline test/gtest/test_basic.cc]: minor: maybe better use `gtest_main.cc` or define a separate file with `main`?

File: test/gtest/test_basic.cc [hunk 4]
```diff
@@ -5,10 +5,285 @@
...
+  };
+  for (auto pair : todo) {
+    mem = create_xpmem(pair.first, pair.second);
+    ASSERT_NE(-1, mem.first);
+    EXPECT_EQ(-1, mem.second);
+    ASSERT_EQ(0, xpmem_remove(mem.first));
+  }
+}
+
+struct xpmem_context {
```
Comment (brminich) [inline test/gtest/test_basic.cc]: maybe better use class for consistency with other tests?

File: test/gtest/test_multi.cc [hunk 5]
```diff
@@ -0,0 +1,615 @@
...
+    ASSERT_NE(-1, apid);
+  }
+
+  void TearDown() {
+    if (data_ptr) {
+      ASSERT_EQ(0, munmap(data_ptr, data_size));
+      data_ptr = nullptr;
+    }
+    test_multi::TearDown();
+  }
```
Comment (brminich) [inline test/gtest/test_multi.cc:112]: pls add empty line  after

File: test/gtest/test_multi.cc [hunk 6]
```diff
@@ -0,0 +1,615 @@
...
+
+        action(att_ptr.get());
+      }
+      Wait();
+    });
+  }
+  testing::AssertionResult test_pattern_check(void* ptr, size_t size, int seed);
+};
+
+TEST_F(test_multi, setup_and_teardown) { ASSERT_TRUE(m_area); }
```
Comment (brminich) [inline test/gtest/test_multi.cc]: i'd not test the test class itself

File: test/gtest/test_multi.cc [hunk 7]
```diff
@@ -0,0 +1,615 @@
...
+  auto fork_action = [&](void* ptr) {
+    // Read fault, but pattern won't be there
+    ASSERT_FALSE(test_pattern_check(ptr, size, seed));
+  };
+  StartAttach(peers, data_ptr, size, fork_action);
+
+  Share()->segid[0] = segid;
+  SignalAll();
+  SignalAll();
+}
```
Comment (brminich) [inline test/gtest/test_multi.cc]: can be a common code

File: test/gtest/test_multi.cc [hunk 8]
```diff
@@ -0,0 +1,615 @@
...
+      seed = pattern_next(seed);
+      data[i] = seed;  // Uses lowest byte only
+    }
+  };
+  StartAttach(peers, data_ptr, size, fork_action);
+
+  WaitForAll();  // Wait for them to be ready
+  Share()->segid[0] = segid;
+  SignalAll();  // Start fork action
+  SignalAll();  // Waits for them
```
Comment (brminich) [inline test/gtest/test_multi.cc]: can we define a func/method which calls `SignallAll` twice and use it everywhere?

File: test/gtest/test_basic.cc (hunk 1 above)
Comment (tvegas1) [reply test/gtest/test_basic.cc]: removed

File: test/gtest/test_basic.cc (hunk 3 above)
Comment (tvegas1) [reply test/gtest/test_basic.cc]: fixed, added `test_main.cc`

File: test/gtest/test_basic.cc (hunk 4 above)
Comment (tvegas1) [reply test/gtest/test_basic.cc]: fixed

File: test/gtest/test_multi.cc (hunk 5 above)
Comment (tvegas1) [reply test/gtest/test_multi.cc:112]: fixed

File: test/gtest/test_basic.cc (hunk 2 above)
Comment (tvegas1) [reply test/gtest/test_basic.cc]: fixed

File: test/gtest/test_multi.cc (hunk 8 above)
Comment (tvegas1) [reply test/gtest/test_multi.cc]: fixed

File: test/gtest/test_multi.cc (hunk 6 above)
Comment (tvegas1) [reply test/gtest/test_multi.cc]: removed

File: test/gtest/test_multi.cc (hunk 7 above)
Comment (tvegas1) [reply test/gtest/test_multi.cc]: fixed

File: test/gtest/test_basic.cc [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+  ASSERT_NE(-1, segid);
+
+  xpmem_apid_t apid = xpmem_get(segid, 9999, XPMEM_PERMIT_MODE, (void*)0);
+  EXPECT_EQ(-1, apid);
+
+  ASSERT_EQ(0, xpmem_remove(segid));
+}
+
+// Missing: Group permission checks
+// Missing: Seems that xpmem_attach() anyways only allows XPMEM_RDWR
```
Comment (brminich) [inline test/gtest/test_basic.cc]: does it  mean `XPMEM_RDONLY` is not possible to use for xpmem_get?

File: test/gtest/test_basic.cc [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+    xpmem_segid_t segid = xpmem_make(NULL, 10 * page_size(), XPMEM_PERMIT_MODE,
+                                     (void*)make_perms);
+    xpmem_apid_t apid = -1;
+    if (segid != -1) {
+      apid = xpmem_get(segid, get_perms, XPMEM_PERMIT_MODE, NULL);
+    }
+    return std::make_pair(segid, apid);
+  };
+
+  auto mem = create_xpmem(0600, XPMEM_RDONLY);
```
Comment (brminich) [inline test/gtest/test_basic.cc]: maybe also check 0400?

File: test/gtest/test_basic.cc [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+  ASSERT_NE(-1, mem.first);
+  EXPECT_NE(-1, mem.second);
+  ASSERT_EQ(0, xpmem_remove(mem.first));
+
+  std::vector<std::pair<uintptr_t, uintptr_t>> todo = {
+    {0200, XPMEM_RDWR},
```
Comment (brminich) [inline test/gtest/test_basic.cc:156]: maybe also check 0400?

File: test/gtest/test_multi.cc [hunk 12]
```diff
@@ -0,0 +1,712 @@
...
+  }
+};
+
+class test_multi_xpmem : public test_multi {
+ protected:
+  handle<xpmem_segid_t> segid;
+  handle<xpmem_apid_t> apid;
+
+  int seed;
+
```
Comment (brminich) [inline test/gtest/test_multi.cc]: minor: no need for these empty lines

File: test/gtest/test_basic.cc (hunk 9 above)
Comment (tvegas1) [reply test/gtest/test_basic.cc]: fixed, removed comment and added corresponding tests for read only

File: test/gtest/test_basic.cc (hunk 10 above)
Comment (tvegas1) [reply test/gtest/test_basic.cc]: fixed, split test and added 0400

File: test/gtest/test_basic.cc (hunk 11 above)
Comment (tvegas1) [reply test/gtest/test_basic.cc:156]: fixed

File: test/gtest/test_multi.cc (hunk 12 above)
Comment (tvegas1) [reply test/gtest/test_multi.cc]: fixed

**Review by brminich — DISMISSED** [review]

File: test/gtest/common/xpmem_test.cc [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+
+int xpmem_test::pattern_next(int seed) {
+  // LFSR
+  return (seed << 1) | (__builtin_parityl(seed & 1337));
+}
+
+std::vector<int> xpmem_test::randomized_sequence(int count) {
+  std::vector<long> key(count);
+  std::vector<int> value(count);
+  std::for_each(key.begin(), key.end(), [](auto &v) { v = random(); });
```
Comment (yosefe) [inline test/gtest/common/xpmem_test.cc]: can we use std::random_shuffle?

File: test/gtest/common/xpmem_test.cc [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+  for (int i = 0; i < value.size(); i++) {
+    value[i] = i;
+  }
+  std::sort(value.begin(), value.end(),
+            [&key](int a, const int b) { return key[a] < key[b]; });
+  return value;
+}
+
+// Return 'count' mmap'd areas of 'size' with holes before and after
+std::vector<std::pair<uint8_t*, size_t>> xpmem_test::mmap_holes(
```
Comment (yosefe) [inline test/gtest/common/xpmem_test.cc]: why not void* like the other mmap() function?

File: test/gtest/common/xpmem_test.cc [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+  }
+  return {};
+}
+
+size_t xpmem_test::mmap_holes_sum(size_t size, int count) {
+  size_t n = size / sizeof(uint32_t) * count;
+  return n * (n + 1) / 2;
+}
+
+void xpmem_test::munmap_holes(std::vector<std::pair<uint8_t*, size_t>> &areas) {
```
Comment (yosefe) [inline test/gtest/common/xpmem_test.cc]: maybe define type for mmap area?

File: test/gtest/common/xpmem_test.h [hunk 16]
```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright (c) NVIDIA CORPORATION & AFFILIATES, 2023. ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
```
Comment (yosefe) [inline test/gtest/common/xpmem_test.h:5]: space line after (in other files as well)

File: test/gtest/common/xpmem_test.h [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+extern "C" {
+#include <xpmem.h>
+}
+#include <sys/mman.h>
+#include <unistd.h>
+
+#include <memory>
+
+template <typename T>
+class handle {
```
Comment (yosefe) [inline test/gtest/common/xpmem_test.h]: maybe can use std::unique_ptr instead?

File: test/gtest/test_multi.cc [hunk 18]
```diff
@@ -0,0 +1,710 @@
...
+  SignalAll();  // Start and wait for peers
+  WaitForAll();
+
+  pid_t pid = fork();
+  EXPECT_LT(-1, pid);
+
+  if (!pid) {
+    for (;;)
+      ;  // busy loop, waiting for kill signal below
```
Comment (yosefe) [inline test/gtest/test_multi.cc]: maybe sleep to not consume CPU?

File: test/gtest/common/xpmem_test.cc (hunk 13 above)
Comment (tvegas1) [reply test/gtest/common/xpmem_test.cc]: fixed, using std::shuffle as std::random_shuffle is on deprecation path.

File: test/gtest/common/xpmem_test.cc (hunk 14 above)
Comment (tvegas1) [reply test/gtest/common/xpmem_test.cc]: fixed, using dedicated class as suggested below (seemed easier for pointer arithmetic).

File: test/gtest/common/xpmem_test.cc (hunk 15 above)
Comment (tvegas1) [reply test/gtest/common/xpmem_test.cc]: fixed, now using specific class to track list of areas.

File: test/gtest/common/xpmem_test.h (hunk 16 above)
Comment (tvegas1) [reply test/gtest/common/xpmem_test.h:5]: fixed once

File: test/gtest/common/xpmem_test.h (hunk 17 above)
Comment (tvegas1) [reply test/gtest/common/xpmem_test.h]: cleaned up handle, now using it through std::unique_ptr.

File: test/gtest/test_multi.cc (hunk 18 above)
Comment (tvegas1) [reply test/gtest/test_multi.cc]: fixed

File: test/gtest/common/xpmem_test.h [hunk 19]
```diff
@@ -20,33 +21,19 @@ extern "C" {
...
-  ~handle() { m_dtor(m_h); }
-
- private:
-  static int empty_dtor(T h) { return 0; }
-  T m_h;
-  dtor_t m_dtor;
+  ~handle() { F(m_h); }
 };
 
+using handle_segid = handle<xpmem_segid_t, xpmem_remove>;
```
Comment (yosefe) [inline test/gtest/common/xpmem_test.h]: can we use std::unique_ptr here as well? and remove handle class altogether?

File: test/gtest/common/xpmem_test.h [hunk 20]
```diff
@@ -91,10 +72,36 @@ class xpmem_test {
...
   static std::vector<int> randomized_sequence(int count);
-  std::vector<std::pair<uint8_t*, size_t>> mmap_holes(size_t hole_size,
-                                                      size_t size, int count,
-                                                      bool fault_pages = true);
-  void munmap_holes(std::vector<std::pair<uint8_t*, size_t>> &areas);
-  size_t mmap_holes_sum(size_t size, int count);
+
+ public:
+  static void *mmap(size_t size, int prot = PROT_READ | PROT_WRITE,
+                    int flags = MAP_PRIVATE | MAP_ANONYMOUS) {
```
Comment (yosefe) [inline test/gtest/common/xpmem_test.h]: maybe define costs
DEFAULT_MMAP_PROT
DEFAULT_MMAP_FLAGS

Comment (tvegas1) [reply test/gtest/common/xpmem_test.h]: added static const in class

File: test/gtest/common/xpmem_test.h (hunk 19 above)
Comment (tvegas1) [reply test/gtest/common/xpmem_test.h]: fixed, but delete class is still needed to provide default constructor for segid and apid class members in test_multi_xpmem.

**Review by yosefe — DISMISSED** [review]

**Review by yosefe — APPROVED** [review]

**Pull request merged**